File: 3-Programacion-Concurrente/Practicas-Resueltas/Tp2-Pasaje-De-Mensajes/entendiendoLosMensajes/sarasa.go
programa eje4
procesos
  proceso limpiarFlores(ES cantFlores: numero)
  comenzar
    mientras(HayFlorEnLaEsquina)
      tomarFlor
      cantFlores:= cantFlores+1
  fin
  proceso limpiarPapeles(ES cantPapeples: numero)
  comenzar
    mientras(HayPapelEnLaEsquina)
      tomarPapel
      cantPapeples:= cantPapeples+1
  fin
  proceso calcularEsquina(ES Av: numero; ES Ca: numero)
  variables
    a1: numero
    a2: numero
  comenzar  
    Random(a1, 25,75)
    Random(a2, 25,75)
    Av:= a1
    Ca:= a2
  fin
  proceso quienEntra(ES botX: numero; ES cantTotal: numero)
  variables
    queRobot: numero
  comenzar
    Random(queRobot, 1,4)
    cantTotal:= 0
    si(queRobot = 1)
      EnviarMensaje(queRobot, bot1)
      EnviarMensaje(0,bot2)
      EnviarMensaje(0,bot3)
      EnviarMensaje(0,bot4) 
      RecibirMensaje(cantTotal, bot1)        
    sino
      si (queRobot= 2)
        EnviarMensaje(queRobot, bot2)
        EnviarMensaje(0, bot1)
        EnviarMensaje(0,bot3)
        EnviarMensaje(0,bot4)
        RecibirMensaje(cantTotal, bot2)
      sino 
        si(queRobot= 3)
          EnviarMensaje(queRobot, bot3)
          EnviarMensaje(0, bot1)
          EnviarMensaje(0,bot2)
          EnviarMensaje(0,bot4)
          RecibirMensaje(cantTotal, bot3)
        sino
          si(queRobot= 4)
            EnviarMensaje(queRobot, bot4)
            EnviarMensaje(0,bot1)
            EnviarMensaje(0,bot2)
            EnviarMensaje(0,bot3)
            RecibirMensaje(cantTotal, bot4)
      botX:= queRobot
  fin
  proceso asignarIds
  comenzar
    EnviarMensaje(1,bot1)
    EnviarMensaje(2,bot2)
    EnviarMensaje(3,bot3)
    EnviarMensaje(4,bot4)
  fin
  proceso inicializarContadores(ES cantBot1: numero;ES cantBot2: numero;ES cantBot3: numero;ES cantBot4: numero)
  comenzar
    cantBot1:= 0
    cantBot2:= 0
    cantBot3:= 0
    cantBot4:= 0
  fin 
  proceso asignacionCantId(E id: numero; E cantTotal: numero; ES cantBot1: numero; ES cantBot2: numero; ES cantBot3: numero; ES cantBot4: numero)
  comenzar
    si (id = 1)
      cantBot1:= cantBot1 + cantTotal
    sino
      si(id = 2)
        cantBot2:= cantBot2 + cantTotal
      sino
        si(id = 3)
          cantBot3:= cantBot3 + cantTotal
        sino
          si(id = 4)
            cantBot4:= cantBot4 + cantTotal
  fin
  proceso sacarMaxCantidad(ES idMax: numero; ES cantBot1: numero; ES cantBot2: numero; ES cantBot3: numero; ES cantBot4: numero)
  variables
    maxCantidad: numero
  comenzar
    idMax := 1
    maxCantidad := cantBot1
    si cantBot2 > maxCantidad
      maxCantidad := cantBot2
      idMax := 2
    si cantBot3 > maxCantidad
      maxCantidad := cantBot3
      idMax := 3
    si cantBot4 > maxCantidad
      maxCantidad := cantBot4
      idMax := 4
  fin
  proceso chequearSiGane(E gane: numero)
  comenzar
    si(gane = 0)      
      Informar('NOGane',0)
    sino
      si(gane = 1)
        Pos(50,50)
        Informar('Gane',1)
  fin
  proceso informarGanador(E id: numero)
  comenzar
    si(id = 1)
      EnviarMensaje(1, bot1)
      EnviarMensaje(0,bot2)
      EnviarMensaje(0,bot3)
      EnviarMensaje(0,bot4)      
    sino
      si (id = 2)
        EnviarMensaje(1, bot2)
        EnviarMensaje(0, bot1)
        EnviarMensaje(0,bot3)
        EnviarMensaje(0,bot4)
      sino 
        si(id = 3)
          EnviarMensaje(1, bot3)
          EnviarMensaje(0, bot1)
          EnviarMensaje(0,bot2)
          EnviarMensaje(0,bot4)
        sino
          si(id = 4)
            EnviarMensaje(1, bot4)
            EnviarMensaje(0,bot1)
            EnviarMensaje(0,bot2)
            EnviarMensaje(0,bot3)
  fin
	proceso logicaNormal
	comenzar
		EnviarMensaje(10,bot1)
		EnviarMensaje(10,bot2)
		EnviarMensaje(10,bot3)
		EnviarMensaje(10,bot4)
	fin
	proceso logicaTermino
	comenzar
		EnviarMensaje(20,bot1)
		EnviarMensaje(20,bot2)
		EnviarMensaje(20,bot3)
		EnviarMensaje(20,bot4)
	fin
areas
  trabajadores: AreaPC(25,25,75,75)
  state: AreaC(1,1,100,1)
  {posGanador: AreaC(50,50,50,50)}
robots
  robot  trabajador
  variables
    cantTotal: numero
    Av: numero
    Ca: numero
    PosInicialAV: numero
    PosInicialCA: numero
    cantFlores: numero
    cantPapeles: numero
    id: numero
    Gane: numero
    ok: numero
    logica: numero
  comenzar
    cantFlores:= 0
    cantPapeles:= 0
    PosInicialAV:= PosAv
    PosInicialCA:= PosCa
    RecibirMensaje(id, botFiscalizador)    {ya tiene cada uno un Id}
		RecibirMensaje(logica, botFiscalizador) {Que LOGICA aplicar  (10 normal) (20 termino}
		si(logica = 10)
			RecibirMensaje(ok, botFiscalizador)	{si este entra o no}
			si (ok <> 0)
				calcularEsquina(Av,Ca)
				Pos(Av,Ca)
				limpiarFlores(cantFlores)
				limpiarPapeles(cantPapeles)
			cantTotal:=  cantFlores + cantPapeles
			EnviarMensaje(cantTotal, botFiscalizador)
			Pos(PosInicialAV, PosInicialCA)
    
    {ESTA PARTE NO VA}
    sino
			si(logica = 20)
				RecibirMensaje(Gane, botFiscalizador) {1, si gane; 0 sino gane}
				chequearSiGane(Gane)
  fin
  robot  fiscalizador
  variables  {aca me serviria un vector contador para los 4 robots}
    id: numero
    cantTotal: numero
    cantBot1: numero
    cantBot2: numero
    cantBot3: numero
    cantBot4: numero
    idMax: numero
  comenzar
    asignarIds
    inicializarContadores(cantBot1,cantBot2,cantBot3,cantBot4)
    repetir 10 
			logicaNormal
      quienEntra(id,cantTotal)  {este se encarga recibir, el id y la cant}
      asignacionCantId(id,cantTotal,cantBot1,cantBot2,cantBot3,cantBot4)
    Informar('Saliste',122)
    sacarMaxCantidad(idMax,cantBot1,cantBot2,cantBot3,cantBot4)
		logicaTermino
    informarGanador(idMax)
  fin
variables
  bot1: trabajador
  bot2: trabajador
  bot3: trabajador
  bot4: trabajador
  botFiscalizador: fiscalizador
comenzar
  {lugar donde se mueven}
  AsignarArea(bot1, trabajadores)
  AsignarArea(bot2, trabajadores)
  AsignarArea(bot3, trabajadores)
  AsignarArea(bot4, trabajadores)
  {LugarDeEspera}
  AsignarArea(bot1, state)
  AsignarArea(bot2, state)
  AsignarArea(bot3, state)
  AsignarArea(bot4, state)
  AsignarArea(botFiscalizador, state)
  {Asignando sus posiciones de espera}
  Iniciar(botFiscalizador, 1,1)
  Iniciar(bot1, 25,1)
  Iniciar(bot2, 30,1)
  Iniciar(bot3, 35,1)
  Iniciar(bot4, 40,1)
fin

File: 3-Programacion-Concurrente/parciales/cursada/2018/sarasa.java
programa arn
procesos
  proceso asignarIds
  comenzar
    EnviarMensaje(1, bot1)
    EnviarMensaje(2, bot2)
    EnviarMensaje(3, bot3)
    EnviarMensaje(4, bot4)
  fin
  proceso calleRandom(ES avRandom: numero; ES caRandom: numero)
  variables
    auxAv, auxCa: numero
  comenzar
    Random(auxAv, 45, 62)
    Random(auxCa, 48, 69)
    avRandom:= auxAv
    caRandom:= auxCa
  fin
  proceso mandarMensajeAlganador(E idMax: numero)  {avisa al que gano}
  comenzar
    si(idMax = 1)
      EnviarMensaje(1, bot1) {1 = ganaste | 2 = perdiste}
      EnviarMensaje(2,bot2)
      EnviarMensaje(2,bot3)
      EnviarMensaje(2,bot4)
    sino
      si(idMax = 2)
        EnviarMensaje(2, bot1) {1 = ganaste | 2 = perdiste}
        EnviarMensaje(1,bot2)
        EnviarMensaje(2,bot3)
        EnviarMensaje(2,bot4)
      sino
        si(idMax = 3)
          EnviarMensaje(2, bot1) {1 = ganaste | 2 = perdiste}
          EnviarMensaje(2,bot2)
          EnviarMensaje(1,bot3)
          EnviarMensaje(2,bot4)
        sino
          EnviarMensaje(2, bot1) {1 = ganaste | 2 = perdiste}
          EnviarMensaje(2,bot2)
          EnviarMensaje(2,bot3)
          EnviarMensaje(1,bot4)
  fin
  proceso  inicializarCantidades(ES bot1Cant: numero;ES bot2Cant: numero;ES bot3Cant: numero;ES bot4Cant: numero)
  comenzar
    bot1Cant:= 0
    bot2Cant:= 0
    bot3Cant:= 0
    bot4Cant:= 0
  fin
  proceso sacarMax(ES idMax: numero; E bot1Cant: numero; E bot2Cant: numero;E bot3Cant: numero; E bot4Cant: numero; ES max: numero)
  comenzar
    max:= bot1Cant
    idMax:= 1
    si(bot2Cant > max)
      max:= bot2Cant
      idMax:= 2
    si(bot3Cant > max)
      max:= bot3Cant
      idMax:= 3
    si(bot4Cant > max)
      max:= bot4Cant
      idMax:= 4
  fin
areas
  areaRecoleccion: AreaPC(45,48,62,69)
  areaBot1: AreaP(10,10,10,10)  {areas default}
  areaBot2: AreaP(11,10,11,10)
  areaBot3: AreaP(12,10,12,10)
  areaBot4: AreaP(13,10,13,10)
  areaJefe: AreaP(1,1,1,1)
robots
  robot  recolector
  variables
    id: numero
    ok: boolean
    avRandom, caRandom: numero
    caIni, avIni: numero
    cantFlores,cantPapeles: numero
    avisoGanador: numero
  comenzar
    RecibirMensaje(id, jefe)
    avIni:= PosAv
    caIni:= PosCa
    cantFlores:= 0
    cantPapeles:= 0
    repetir 5
      calleRandom(avRandom, caRandom)
      {Random(avRandom, 45, 62)}
     {Random(caRandom, 48, 69)}
      EnviarMensaje(id, jefe)
      BloquearEsquina(avRandom, caRandom)
      Pos(avRandom, caRandom)
      mientras(HayFlorEnLaEsquina) | (HayPapelEnLaEsquina)
        ok:= V
        si(HayFlorEnLaEsquina)
          tomarFlor
          cantFlores:= cantFlores+1
          EnviarMensaje(ok, jefe)
        si(HayPapelEnLaEsquina)
          tomarPapel
          cantPapeles:= cantPapeles+1
          EnviarMensaje(ok, jefe)
      ok:= F
      EnviarMensaje(ok, jefe)
      Pos(avIni,caIni)
      LiberarEsquina(avRandom, caRandom)
    EnviarMensaje(0, jefe)               {le aviso que un robot ya termino}
    RecibirMensaje(avisoGanador, jefe)  {avisa si gano}
    si(avisoGanador = 1)
      Informar('Gane',231)
      repetir cantFlores
        depositarFlor
      repetir cantPapeles
        depositarPapel
  fin
  robot  coordinador
  variables
    avRandom, caRandom: numero
    ok: boolean
    contOk: numero
    bot1Cant, bot2Cant, bot3Cant, bot4Cant: numero
    cont4,max,idMax: numero
    id: numero
  comenzar
    asignarIds
    cont4:= 0
    inicializarCantidades(bot1Cant, bot2Cant, bot3Cant, bot4Cant)
    mientras(cont4 < 4)
      RecibirMensaje(id, *)
      si(id = 1)
        RecibirMensaje(ok, bot1)
        mientras(ok)
          bot1Cant:= bot1Cant+1
          RecibirMensaje(ok, bot1)
      sino
        si(id = 2)
          RecibirMensaje(ok, bot2)
          mientras(ok)
            bot2Cant:= bot2Cant+1
            RecibirMensaje(ok, bot2)
        sino
          si(id = 3)
            RecibirMensaje(ok, bot3)
            mientras(ok)
              bot3Cant:= bot3Cant+1
              RecibirMensaje(ok, bot3)
          sino  {es 4}
            si(id = 4)
              RecibirMensaje(ok, bot4)
              mientras(ok)
                bot4Cant:= bot4Cant+1
                RecibirMensaje(ok, bot4)
            sino
              cont4:= cont4 +1
    sacarMax(idMax,bot1Cant,bot2Cant,bot3Cant,bot4Cant,max)
    mandarMensajeAlganador(idMax)
  fin
variables
  bot1, bot2, bot3, bot4: recolector
  jefe: coordinador
comenzar
  AsignarArea(bot1, areaBot1)  {area default}
  AsignarArea(bot2, areaBot2)
  AsignarArea(bot3, areaBot3)
  AsignarArea(bot4, areaBot4)
  AsignarArea(jefe, areaJefe)
  AsignarArea(bot1, areaRecoleccion)  {area en comun}
  AsignarArea(bot2, areaRecoleccion)
  AsignarArea(bot3, areaRecoleccion)
  AsignarArea(bot4, areaRecoleccion)
  Iniciar(bot1, 10, 10)
  Iniciar(bot2, 11, 10)
  Iniciar(bot3, 12, 10)
  Iniciar(bot4, 13, 10)
  Iniciar(jefe, 1, 1)
fin